### PR TITLE
Fix autogrow width issue

### LINKED ISSUE #4372: Autogrow width issue
Hi,
I am trying to use the Autogrow plugin, I set the height "CKEDITOR.config.height = 150;" and width "CKEDITOR.config.width = '600';". Then I fill with the content so that the height matches, at this point the whole editor changes the width to a smaller one. Each new line reduces the width by 2px.

### PR DESCRIPTION
**Pull request by sculpt0r**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [X] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4372](https://github.com/ckeditor/ckeditor4/issues/4372): Fixed: [Autogrow](https://ckeditor.com/cke4/addon/autogrow) plugin changes editor's width when used with absolute [`config.width`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-width) value. 
```

## What changes did you make?

*Give an overview…*
Add tests for `autogrow` plugin with multiple width cases.
Change creation of `width` value for resizing. Based on this logic ([#4372 (comment)](https://github.com/ckeditor/ckeditor4/issues/4372#issuecomment-731962692))

## Which issues does your PR resolve?

Closes #4372

### CODE REVIEW

File: plugins/autogrow/plugin.js [hunk 1]
```diff
@@ -147,7 +147,13 @@
 			// to the one set by previous resizeEditor() call.
 			if ( newHeight != currentHeight && lastHeight != newHeight ) {
 				newHeight = editor.fire( 'autoGrow', { currentHeight: currentHeight, newHeight: newHeight } ).newHeight;
-				editor.resize( editor.container.getStyle( 'width' ), newHeight, true );
+
+				var boxSizingType = editor.container.getComputedStyle( 'box-sizing' );
+				var isBorderBox = boxSizingType === 'border-box';
+
+				var width = editor.container.getSize( 'width', isBorderBox );
```
Comment (jacekbogdanski) [inline plugins/autogrow/plugin.js:154]: We are declaring variables using a single `var` statement when possible. Please, join variables.

File: tests/plugins/autogrow/autogrow.js [hunk 2]
```diff
@@ -8,7 +8,58 @@
 
 	bender.editor = {};
 
+	bender.createEditor = function( width, paragraphs ) {
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/autogrow.js:11]: This method does much more than creating an editor. It set up the whole testing environment and does assertions.

File: tests/plugins/autogrow/autogrow.js [hunk 3]
```diff
@@ -8,7 +8,58 @@
...
+			bender.createEditor( 200, 6 );
+		},
+		// (#4372)
+		'test autogrow for editor width 0': function() {
+			bender.createEditor( 0, 6 );
+		},
+		// (#4372)
+		'test autogrow for editor width auto': function() {
+			bender.createEditor( 'auto', 6 );
+		},
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/autogrow.js:62]: Since we are not changing the number of paragraphs anywhere, maybe let's just hardcode it inside the test method? We could also choose a bit higher value to make sure to not depend on some edge case, like 10.

File: tests/plugins/autogrow/autogrow.js [hunk 4]
```diff
@@ -8,7 +8,58 @@
...
+			config: {
+				width: width
+			}
+		}, function( bot ) {
+			bot.setData( autogrowTools.getTestContent( paragraphs ), function() {
+				var editor = bot.editor;
+				var initialEditorSize = autogrowTools.getEditorSize( editor );
+
+				editor.once( 'afterCommandExec', function() {
+					resume( function name( params ) {
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/autogrow.js:23]: `params` is not used, so could be omitted. We could skip`name` name also since it's not telling anything and we are using anonymous functions in most cases.

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
Please, add a manual test for the changes.

**Comment by sculpt0r** [comment]
> Please, add a manual test for the changes.

Added manual test for this bug.
Should I also edit PR to check 'Manual tests'?

File: tests/plugins/autogrow/_helpers/tools.js [hunk 5]
```diff
@@ -19,8 +19,36 @@ var autogrowTools = ( function() {
 		return html;
 	}
 
+	function testEditorSizeWithContent( editorWidth ) {
+		bender.editorBot.create( {
+			name: 'EditorWidth' + editorWidth,
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/_helpers/tools.js:24]: Running this test twice with the same `editorWidth` will result in an error as we will try to create two editors with the same name. Since this test helper has been extracted as a separate tool method, maybe we could improve it by adding some more random editor name? As an example. you could use `new Date().getTime()` as it's really unlikely it will match milliseconds:

```js
name: 'editor_' + new Date().getTime()
```

File: tests/plugins/autogrow/manual/4372.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
+@bender-tags: 4.15.1, bug, 7372
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, autogrow
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/manual/4372.md:3]: Let's add some minimal editor plugins, so the editor toolbar loads nicely.
```suggestion
@bender-ckeditor-plugins: wysiwygarea, autogrow, basicstyles, toolbar, resize
```

File: tests/plugins/autogrow/manual/4372.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+
+1. Focus inside editable window.
+2. Add or remove extra lines to force editor resize.
+3. Observe editor width.
+
+#### Expected result:
+The editor should have constant width. Only height is changing according to vertical amount of content.
+
+#### Unexpected result:
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/manual/4372.md:13]:
```suggestion
## Unexpected
```

File: tests/plugins/autogrow/manual/4372.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+@bender-tags: 4.15.1, bug, 7372
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, autogrow
+
+
+1. Focus inside editable window.
+2. Add or remove extra lines to force editor resize.
+3. Observe editor width.
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/manual/4372.md:8]:
```suggestion
1. Focus the editor.
2. Press `enter` multiple times to force the editor to resize.
```

File: tests/plugins/autogrow/manual/4372.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+#### Expected result:
+The editor should have constant width. Only height is changing according to vertical amount of content.
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/manual/4372.md:11]:
```suggestion
Editor width is not changing during auto-resize.
```

File: tests/plugins/autogrow/manual/4372.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+#### Unexpected result:
+With each vertical resize editor is shrinking for about margins + paddings.
```
Comment (jacekbogdanski) [inline tests/plugins/autogrow/manual/4372.md:14]:
```suggestion
With each vertical resize editor is shrinking for about 2 pixels.
```

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
Just some minor changes required.

File: tests/plugins/autogrow/manual/4372.md (hunk 10 above)
Comment (sculpt0r) [reply tests/plugins/autogrow/manual/4372.md:14]: That was my first thought. Later it comes to my mind (I may be over-obsessed) if the default margin and padding will change, those `2px` will be invalid. Anyway, I apply suggestion :)

File: tests/plugins/autogrow/manual/4372.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+@bender-tags: 4.15.1, bug, 7372
```
[the remark this answers is not in the document]
Comment (sculpt0r) [reply tests/plugins/autogrow/manual/4372.md:1]: I'm confused about one thing:
> We are at 4.15.1. version already. The upcoming release version for bugs is 4.15.2.

So I should tag it with a future version of the editor?

>7372 tag is invalid here. It should be the ticket number, so 4372 in our case.

:speak_no_evil: I mean it, so shame on me for this! :(

[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/autogrow/manual/4372.md]: Yes, it should reflect the target release, not the current one, as we are not providing backward patches.

**Review by jacekbogdanski — APPROVED** [review]
Looks good!

I've ignored unit tests on IE8 as they were failing for no good reason despite that the manual tests worked fine.

**Comment by sculpt0r** [comment]
> Looks good!
> 
> I've ignored unit tests on IE8 as they were failing for no good reason despite that the manual tests worked fine.

That `IE` thing... I have to remember to run tests on supported browsers. Thank you.
